lib/builders/latexmk.js

```diff
@@ -75,12 +78,42 @@ export default class LatexmkBuilder extends Builder {
       args.push('-synctex=1')
     }
 
-    if (engineFromMagic) {
-      args.push(`-pdflatex="${engineFromMagic}"`)
-    } else if (customEngine) {
-      args.push(`-pdflatex="${customEngine}"`)
-    } else if (engine && engine !== 'pdflatex') {
-      args.push(`-${engine}`)
+    if (engine.match(latexRegExp)) {
+      args.push(`-latex="${engine}"`)
+      if (outputFormat === 'pdf') {
+        switch (producer) {
+          case 'default':
+            // default is an option to protect a user's .latexmkrc from override
+            args.push('-pdfdvi')
+            break
+          case 'dvipdf':
+            args.push('-pdfdvi')
+            args.push('-e "\\$dvipdf = \'dvipdf %O %S %D\';"')
+            break
+          case 'dvipdfm':
+            args.push('-pdfdvi')
+            args.push('-e "\\$dvipdf = \'dvipdfm %O -o %D %S\';"')
+            break
+          case 'dvipdfmx':
+            args.push('-pdfdvi')
+            args.push('-e "\\$dvipdf = \'dvipdfmx %O -o %D %S\';"')
+            break
+          case 'ps2pdf':
+            args.push(`-pdfps`)
+            break
+          default:
+            atom.notifications.addWarning(`Unknown PDF producer ${producer}, using fallback dvipdf.`)
+            args.push('-pdfdvi')
+            break
+        }
+      } else {
+        args.push(`-${outputFormat}`)
+      }
+    } else {
+      if (engine !== 'pdflatex') {
+        args.push(`-pdflatex="${engine}"`)
+      }
+      args.push(`-${outputFormat}`)
```

Comment:
With these changes `constructArgs` is doing way too much. We need to extract some more granular logic. It's becoming really hard to read. Try to split into logical units... one example might be something like `getEngineArgs` which can also be split into smaller units (possibly one like `getPdfProducerArgs`).